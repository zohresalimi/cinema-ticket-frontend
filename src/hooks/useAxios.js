import { useEffect, useReducer, useMemo, useCallback } from "react";
import axios from "axios";
import qs from "qs";

const baseUrl = process.env.REACT_APP_BASE_URL;
const actions = {
  REQUEST_START: "REQUEST_START",
  REQUEST_END: "REQUEST_END",
};

async function fetchData({ path, ...config }, dispatch) {
  try {
    dispatch({ type: actions.REQUEST_START });
    const result = await axios({
      ...config,
      url: `${baseUrl}${path}`,
      paramsSerializer: (params) => {
        return qs.stringify(params);
      },
    });
    dispatch({ type: actions.REQUEST_END, payload: result.data });
  } catch (err) {
    dispatch({
      type: actions.REQUEST_END,
      payload: err,
      error: true,
    });
  }
}

const useAxios = (path, config) => {
  function reducer(state, action) {
    switch (action.type) {
      case actions.REQUEST_START:
        return {
          ...state,
          response: null,
          loading: true,
          error: null,
        };
      case actions.REQUEST_END:
        return {
          ...state,
          loading: false,
          ...(action.error ? {} : { data: action.payload.data }),
          [action.error ? "error" : "response"]: action.payload,
        };
      default:
        return state;
    }
  }

  function createInitialState({ manual }) {
    return {
      loading: !manual,
      error: null,
      response: null,
    };
  }

  const options = useMemo(
    () => ({
      data: {},
      method: "get",
      manual: false,
      path,
      ...config,
    }),
    [config, path]
  );
  const [state, dispatch] = useReducer(reducer, createInitialState(options));

  useEffect(() => {
    if (!options.manual) {
      fetchData(options, dispatch);
    }
  }, [options, dispatch, path]);

  const refetch = useCallback(
    (payload) => {
      return fetchData({ ...options, ...payload }, dispatch);
    },
    [options, dispatch]
  );

  return [state, refetch];
};

export default useAxios;
