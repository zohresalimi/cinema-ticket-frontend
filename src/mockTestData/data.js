import React from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";

export const getTestStore = () => {
  return {
    dispatch: () => {},
    state: {
      user: {
        name: "",
        email: "",
      },
      movies: {
        children: [
          {
            _id: "5fa71823adc463397a50c501",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "trolls 2: world tour",
            originalTitle: "swedish",
            premiere: "2020-10-03T15:18:52.722Z",
            rooms: ["5fa653d0a7f1201999dd1ac0"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        premiered: [
          {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        upcoming: [
          {
            _id: "5fa71823adc463397a50c503",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "honest thief",
            originalTitle: "swedish",
            premiere: "2021-01-03T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        allMovies: {
          "5fa71823adc463397a50c501": "children",
          "5fa71823adc463397a50c502": "premiered",
          "5fa71823adc463397a50c503": "upcoming",
        },
      },
      cinemas: [
        {
          _id: "5fa64fe234fbbf9945f3a3ce",
          name: "filmstaden kista",
          purchaseEndTime: "21:30",
          purchaseStartTime: "8:00",
          rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
        },
      ],
      ticket: {
        movie: {},
        status: false,
        quantity: 0,
        price: 0,
        seatNumbers: new Map(),
        user: {},
        showing: {},
      },
      showings: [
        {
          _id: "5fbe9bd31f58462b5c759b5c",
          capacity: 4,
          cinema: "5fa64fe234fbbf9945f3a3ce",
          endTime: null,
          movie: {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
          price: 100,
          room: { _id: "5fa653c0643c06586bba7d78" },

          seats: [
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
          ],
          startTime: "2020-11-13T16:00:00.000Z",
        },
      ],
    },
  };
};

export const WithProvider = (props) => {
  const { state, dispatch } = getTestStore();
  return (
    <Theme>
      <AppContext.Provider
        value={{ state: props.defaultValue || state, dispatch }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
