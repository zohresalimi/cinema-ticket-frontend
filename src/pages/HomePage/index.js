import React, { useEffect, useContext } from "react";
import useAxios from "../../hooks/useAxios";
import AppContext from "../../store/context";
import TopBar from "../../components/TopBar";
import FeaturedMovie from "../../components/FeaturedMovie";
import Slider from "../../components/Slider";
import { SET_CURRENT_MOVIE_REDUCER } from "../../constants";

function HomePage() {
  const { state, dispatch } = useContext(AppContext);
  const { currentMovies } = state;
  const [{ response }] = useAxios(`/api/v1/movies/current-movie`);

  useEffect(() => {
    if (response) {
      console.log(response.data);
      dispatch({ type: SET_CURRENT_MOVIE_REDUCER, data: response.data });
    }
  }, [response, dispatch]);
  return (
    <>
      <TopBar></TopBar>
      <FeaturedMovie></FeaturedMovie>
      {currentMovies && <Slider items={currentMovies} />}
    </>
  );
}

export default HomePage;
