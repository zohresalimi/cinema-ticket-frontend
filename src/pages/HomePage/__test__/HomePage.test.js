import * as React from "react";
import { when } from "jest-when";
import { render, act, waitFor } from "@testing-library/react";
import * as hooks from "../../../hooks/useAxios";
import "@testing-library/jest-dom/extend-expect";
import { getTestStore, WithProvider } from "../../../mockTestData/data";
import HomePage from "../";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (k) => k,
    i18n: {
      language: "sv",
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

async function renderWrapper(setDefaultState) {
  let component;
  const defaultValue = setDefaultState
    ? {
        user: {},
        movies: {
          upcoming: [],
          children: [],
          premiered: [],
          allMovies: {},
        },
        cinemas: [],
        showings: [],
        ticket: {},
      }
    : null;

  act(() => {
    component = render(
      <WithProvider defaultValue={defaultValue}>
        <HomePage path="/" />
      </WithProvider>
    );
  });

  return component;
}

jest.mock("react-owl-carousel", () => ({ children }) => <div>{children}</div>);
jest.mock("../../../components/TopBar", () => () => (
  <div>TopBar Component</div>
));
jest.mock("../../../components/FeaturedMovie", () => () => (
  <div>Featured Movie Component</div>
));
jest.mock("../../../components/Cards", () => ({ item }) => (
  <div data-testid="card-item">{item.name}</div>
));

describe("Home page Component Testing", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  test("take snapshot", async () => {
    const { container } = await renderWrapper(true);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("take snapshot when state has default value", async () => {
    const { container } = await renderWrapper();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should read data from state", async () => {
    const useRefMock = jest.fn(() => ({ current: 1 }));
    jest.spyOn(React, "useRef").mockImplementation(useRefMock);
    const testState = getTestStore().state;
    const mockUseAxios = jest.fn();
    when(mockUseAxios)
      .calledWith(`/api/v1/movies/current-movie`)
      .mockReturnValue([
        {
          response: testState.movies.premiered,
        },
      ]);
    when(mockUseAxios)
      .calledWith(`/api/v1/movies/upcoming`)
      .mockReturnValue([
        {
          response: testState.movies.upcoming,
        },
      ]);
    when(mockUseAxios)
      .calledWith(`/api/v1/movies/by-genre`, 1)
      .mockReturnValue([
        {
          response: testState.movies.children,
        },
      ]);

    jest.spyOn(hooks, "default").mockImplementation(mockUseAxios);

    const { container } = await renderWrapper(true);
    await waitFor(() => expect(mockUseAxios).toHaveBeenCalledTimes(3));
    expect(container.firstChild).toMatchSnapshot();
  });
});
