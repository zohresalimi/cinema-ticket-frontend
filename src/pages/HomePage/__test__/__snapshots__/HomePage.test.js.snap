// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page Component Testing should read data from state 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    TopBar Component
  </div>
  <div>
    Featured Movie Component
  </div>
  <div
    class="black-bg"
  >
    <div
      class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
    >
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          at the cinema now
        </h2>
        <div />
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          upcoming movies
        </h2>
        <div />
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          children and family
        </h2>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`Home page Component Testing take snapshot 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    TopBar Component
  </div>
  <div>
    Featured Movie Component
  </div>
  <div
    class="black-bg"
  >
    <div
      class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
    >
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          at the cinema now
        </h2>
        <div />
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          upcoming movies
        </h2>
        <div />
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          children and family
        </h2>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`Home page Component Testing take snapshot when state has default value 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    TopBar Component
  </div>
  <div>
    Featured Movie Component
  </div>
  <div
    class="black-bg"
  >
    <div
      class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
    >
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          at the cinema now
        </h2>
        <div>
          <a
            href="/premiered/movie-detail/5fa71823adc463397a50c502"
          >
            <div
              data-testid="card-item"
            >
              follow me
            </div>
          </a>
        </div>
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          upcoming movies
        </h2>
        <div>
          <a
            href="/upcoming/movie-detail/5fa71823adc463397a50c503"
          >
            <div
              data-testid="card-item"
            >
              honest thief
            </div>
          </a>
        </div>
      </div>
      <div
        class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
      >
        <h2
          class="movie-title"
        >
          children and family
        </h2>
        <div>
          <a
            href="/children/movie-detail/5fa71823adc463397a50c501"
          >
            <div
              data-testid="card-item"
            >
              trolls 2: world tour
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
